pkg/binder/cache/cache.go

```diff
@@ -17,959 +17,197 @@ limitations under the License.
 
-// createImageStateSummary returns a summarizing snapshot of the given image's state.
-func (cache *binderCache) createImageStateSummary(state *imageState) *framework.ImageStateSummary {
-	return &framework.ImageStateSummary{
-		Size:     state.size,
-		NumNodes: len(state.nodes),
-	}
-}
+	handler.SetPodOpFunc(bc.podOp)
 
-func newBinderCache(ttl, period time.Duration, stop <-chan struct{}, binderName string) *binderCache {
-	bc := &binderCache{
-		binderName: binderName,
-		ttl:        ttl,
-		period:     period,
-		stop:       stop,
-
-		assumedPods:           make(map[string]bool),
-		podsMarkedToBeDeleted: make(map[string]bool),
-		podStates:             make(map[string]*podState),
-		nodeInfoMap:           make(map[string]framework.NodeInfo),
-		imageStates:           make(map[string]*imageState),
-		podGroupPodMap:        make(map[string]map[string]*v1.Pod),
-		podGroupInfo:          make(map[string]*schedulingv1a1.PodGroup),
-		unitStatus:            binderutils.NewUnitStatusMap(),
-		pdbItems:              make(map[string]*pdbItem),
-	}
 	return bc
 }
 
-// PodCount returns the number of pods in the cache (including those from deleted nodes).
-// DO NOT use outside of tests.
-func (cache *binderCache) PodCount() (int, error) {
-	cache.mu.RLock()
-	defer cache.mu.RUnlock()
-	count := 0
-	for _, n := range cache.nodeInfoMap {
-		count += n.NumPods()
-	}
-	return count, nil
-}
-
-func (cache *binderCache) GetPodGroupPods(podGroupName string) []*v1.Pod {
-	cache.mu.RLock()
-	defer cache.mu.RUnlock()
-
-	var pods []*v1.Pod
-	for _, pod := range cache.podGroupPodMap[podGroupName] {
-		pods = append(pods, pod)
-	}
-
-	return pods
+func (cache *binderCache) podOp(pod *v1.Pod, isAdd bool, skippedStores sets.String) error {
+	return cache.CommonStoresSwitch.Range(
+		func(cs commonstore.Store) error {
+			if skippedStores.Len() > 0 && skippedStores.Has(string(cs.Name())) {
+				return nil
+			}
+			if isAdd {
+				return cs.AddPod(pod)
+			}
+			return cs.DeletePod(pod)
+		})
 }
 
-func (cache *binderCache) GetPodGroupInfo(podGroupName string) (*schedulingv1a1.PodGroup, error) {
-	cache.mu.RLock()
-	defer cache.mu.RUnlock()
+func (cache *binderCache) run() {
+	// TODO: Unify and organize metrics related information.
+	go wait.Until(func() {
+		cache.updateMetrics()
+	}, cache.handler.Period(), cache.handler.StopCh())
 
-	pg, found := cache.podGroupInfo[podGroupName]
-	if !found {
-		return nil, errors.New("not found")
-	}
-	return pg, nil
+	cache.CommonStoresSwitch.Range(func(cs commonstore.Store) error {
+		cs.PeriodWorker(cache.handler.Mutex())
+		return nil
+	})
 }
 
-func (cache *binderCache) AssumePod(pod *v1.Pod) error {
-	key, err := framework.GetPodKey(pod)
-	if err != nil {
-		return err
-	}
-
+// TODO: Extend the metrics related.
+// updateMetrics updates cache size metric values for pods, assumed pods, and nodes
+func (cache *binderCache) updateMetrics() {
 	cache.mu.Lock()
```

Comment:
Just RLock、RUnlock is ok，since there no update operation for PodStore or NodeStore

Reply:
There's only read operation for PodStore or NodeStore; however, there's update on binder cache size. If we use read lock, multiple threads can call updateMetrics() and modify the only binder cache size. Thoughts?

Reply:
I also preferto keep the use of write lock.